wasm/jit/jit_amd64.go

```diff
@@ -4594,66 +4569,695 @@ func (c *amd64Compiler) allocateRegister(t generalPurposeRegisterType) (reg int1
+	}
+
+	// Grow the stack.
+	if err := c.callGoFunction(jitCallStatusCodeCallBuiltInFunction, builtinFunctionAddressGrowCallFrameStack); err != nil {
+		return err
+	}
+
+	// After the function call, we have to initialize the stack base pointer and memory reserved registers.
+	c.initializeReservedStackBasePointer()
+	c.initializeReservedMemoryPointer()
+
+	// For call_indirect, we need to push the value back to the register.
+	if !isNilRegister(addrReg) {
+		savedOffsetLocation := c.locationStack.pop()
+		savedOffsetLocation.setRegister(addrReg)
+		c.moveStackToRegister(savedOffsetLocation)
+	}
+
+	// Also we have to re-read the call frame stack pointer into callFrameStackPointerRegister.
+	c.getCallFrameStackPointer(callFrameStackPointerRegister)
+
+	// Now that callframe stack is enough length, we are ready to create a new call frame
+	// for the function call we are about to make.
+	getCallFrameStackFirstItemAddress := c.newProg()
+	jmpIfNotCallFrameStackNeedsGrow.To.SetTarget(getCallFrameStackFirstItemAddress)
+	getCallFrameStackFirstItemAddress.As = x86.AMOVQ
+	getCallFrameStackFirstItemAddress.To.Type = obj.TYPE_REG
+	getCallFrameStackFirstItemAddress.To.Reg = tmpRegister
+	getCallFrameStackFirstItemAddress.From.Type = obj.TYPE_MEM
+	getCallFrameStackFirstItemAddress.From.Reg = reservedRegisterForEngine
+	getCallFrameStackFirstItemAddress.From.Offset = engineGlobalContextCallFrameStackFirstItemAddressOffset
+	c.addInstruction(getCallFrameStackFirstItemAddress)
+
+	// Since call frame stack pointer is the index for engine.callFrameStack slice,
+	// here we get the actual offset in bytes via shifting callFrameStackPointerRegister by callFrameDataSizeMostSignificantSetBit.
+	// That is valid because the size of callFrame struct is a power of 2 (see TestVeifyOffsetValue), which means
+	// multiplying withe the size of struct equals shifting by its most significant bit.
+	shiftCallFrameStackPointer := c.newProg()
+	shiftCallFrameStackPointer.As = x86.ASHLQ
+	shiftCallFrameStackPointer.To.Type = obj.TYPE_REG
+	shiftCallFrameStackPointer.To.Reg = callFrameStackPointerRegister
+	shiftCallFrameStackPointer.From.Type = obj.TYPE_CONST
+	shiftCallFrameStackPointer.From.Offset = int64(callFrameDataSizeMostSignificantSetBit)
+	c.addInstruction(shiftCallFrameStackPointer)
+
+	// At this point, callFrameStackPointerRegister holds the offset in call frame slice in bytes,
+	// and tmpRegister holds the absolute address of the first item of call frame slice.
+	// To illustrate the situation:
+	//
+	//  tmpRegister (holding the absolute address of &callFrame[0])
+	//      |
+	//      [ra.0, rb.0, rc.0, _, ra.1, rb.1, rc.1, _, ra.next, rb.next, rc.next, ...]  <--- call frame stack's data region (somewhere in the memory)
+	//      |                                        |
+	//      <---------------------------------------->
+	//          callFrameStackPointerRegister (holding the offset from &callFrame[0] in bytes.)
```

Comment:
the diagram is ambiguous as it shows an arrow at both ends, which makes it confusing what callFrameStackPointerRegister is actually pointing to. use another LF if needed and you want to show the range and also what the pointer is at